schemachanger: logictest with new schema changer

fqazi opened the pull request:
This pull request will do a large number of small fixes to allow us to enable the logicttest for the declarative schema changer. It's best to look at this pull request commit by commit, since each fix addresses some part limitation of the declarative schema changer.

fqazi committed:
sql: add logictest mode for declarative schema changer

Previously, the declarative schema change only had a single
logictest called new_schema_changer for validating behavior,
so tests would be moved from the existing suite to provide coverage
for newer features in the declarative schema changer.
This was inadequate because it meant that we were missing tons
of coverage from existing tests for schema changes. To address this,
this patch adds a new logic test mode for the declarative schema
changer.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73493)
<!-- Reviewable:end -->

fqazi commented:
Still a number of mismatches @ajwerner :

1. Event logging for drop schema doesn't fully line up
2. Some error messages like DROP TYPE and a few others still need updating
3. Need to filter out and update tests looking for specific jobs
4. There is still a mystery hang inside logic test suite with the declarative schema changer (I think due to transaction retries). I wonder if it's still schema_changer_retry however I can't hit it locally.
5. There might be others, but that's from my quick look

fqazi committed:
schemachanger: prevent dropping of virtual, temporary, and public schemas

Previously, the declarative schema changer incorrectly allowed dropping
virtual, public and temporary schemas, failing within the execution
phase. This patch blocks these operations during the build phase.

Release note: None

fqazi committed:
schemachanger: validate permissions for cascaded drops of views/tables

Previously, the declarative schema changer executed cascaded
drops in a way where permissions were not properly checked dependent
views / tables. As a result it was possible for users to drop objects
they should not have been allowed to, for example a view could
be dropped as dependent even if permissions should not have been
there. To address this, this patch validates if a drop should
be allowed during cascaded operations.

Release note: None

fqazi committed:
schemahchanger: avoid duplicate type reference elements for expressions

Previously, it was possible to have repeat type references added for
expressions as we processed the entire type closure. This patch
ensures that a type reference is only added once per each type.

Release note: None

fqazi committed:
schemachanger: incorrect error when modifying table record types

Previously, the declarative schema changer generated the incorrect
error when attempting to modify implicit record types. To address
this, this patch will generate the same error as the legacy
schema changer.

Release note: None

fqazi committed:
schemachanger: clean up multiple DependedOnBy entries during drop

Previously, the clean up logic for dropping views did not
correctly clean up if multiple dependencies existed inside
a table. For example its valid for entries to exist for
different sets of columns. As a result we could run into
validation errors due to dangling references. To address this,
this patch allows support for cleaning up multiple DepndedOnBy
for a single table.

Release note: None

fqazi committed:
logictest: ignore order of cascaded view drops

The deterministic order for the views listed in event logs
related to cascaded drops varies between the declarative and
legacy schema changer. This patch updates drop_view test to accept
either order.

Release note: None

fqazi committed:
logictest: filter the out the new_scheama_changer setting

Previously, a number of logictests list non-default settings,
as a part of enabling testing for the declarative schema changer
these tests now show the setting to enable it. To address this,
this patch will mask that setting out.

Release note: None

fqazi committed:
schemachanger: type handle cascaded drops properly

Previously, in the declarative schema changer it was possible
to drop types without cleaning up all the references. This could
leading to dangling references / uses of types in tables/views.
To address this, this patch properly blocks these types of drops
similar to the legacy schema changer.

Release note: None

fqazi committed:
schemachanger: missing table name in relationship dependency errors

Previously, the error for dependent table names was incorrectly missing
the table. To address this, this patch will add in the table name into
these dependent object error messages.

Release note: None

fqazi committed:
schemachanger: object list was incorrect for drop database event logs

Previously, when dropping databases the event log entires generated
listen the schemas dropped in the declarative schema changed. This was
incorrect because the legacy schema changer listed the views/tables/sequences
being dropped. To address this, this patch changes the event log entries
generated by the declarative schema changer to list objects instead. Some
tests are also modified since the order of objects generated between
the declarative and legacy schema changer are different.

Release note: None

fqazi committed:
logictest: fixes for declarative schema changer

This commit will do a number of fixes for logictests running
under the declarative schema changer:

1) Add support for the new authorization messages from
   the declarative schema changer.
2) Update test cases to expected different jobs based on
   declarative or legacy schema changer.
3) Exclude experimental_use_new_schema_changer from tests
   checking settings.

Release note: None

fqazi committed:
schemachanger: incorrect message for DROP DATABASE with safe updates

Previously, the declarative schema changer allowed dropping of
the current database when safe updates was enabled. This is
expected to generate an error message. This patch generates
the correct error message for this scenario.

Release Note: None

fqazi committed:
schemachanger: schema owner should be able to drop relations

Previously, you had to be the owner of the relation or database
admin for drops to be allowed. This was incorrect because a schema
owner should be able to also drop these objects. To address this,
this patch fixes the ownership checks to treat the schema owner
as the same for the declarative schema changer.

Release note: None

fqazi committed:
schemachanger: incorrect error generated when modifying virtual tables

Previously, the declarative schema changer did not return the correct
error when trying to modify virtual tables. In some cases run time errors
could be hit. This patch, generates the correct error attempting to modify
virtual tables.

Release note: None

fqazi committed:
schemachanger: check schema change feature flags for declarative

Previously, the declarative schema changer did not enforce
the feature flags. This prevent the schema changes from being
disabled inside the declarative schema changer. To address this,
this patch adds automated checks against the feature flags based
on the statement (taking advantage of the statement tag).

Release Note: None

postamar reviewed (changes requested):
> Reviewed 1 of 1 files at r1, 1 of 2 files at r3, 1 of 1 files at r4, 7 of 8 files at r6, 1 of 1 files at r8, 2 of 3 files at r10, 5 of 10 files at r11, 1 of 2 files at r12, 7 of 9 files at r13, 1 of 1 files at r16, 12 of 12 files at r17, 2 of 2 files at r18, 1 of 1 files at r19, 12 of 12 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73493)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @msbutler)

___
*[pkg/sql/schemachanger/scbuild/descriptor_reader.go, line 39 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-Mtt_GCuBvwAPu74BHG6:-Mtt_GCuBvwAPu74BHG7:b-zhcmny) ([raw file](https://github.com/cockroachdb/cockroach/blob/922e4986aeb2df6b3ddfd135108852eaa3584b9a/pkg/sql/schemachanger/scbuild/descriptor_reader.go#L39)):*
> ```Go
> 	}
> 	return schema
> }
> ```

It seems strange to remove this. 
___
*[pkg/sql/schemachanger/scbuild/name_resolver.go, line 123 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-Mtt_Tq70BNY_DKH4vPV:-Mtt_Tq70BNY_DKH4vPW:br9r0a2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/name_resolver.go#L123)):*
> ```Go
> 	}
> 	// If we own the schema then we can manipulate the underlying relation.
> 	_, schema := b.CatalogReader().MayResolveSchema(b, prefix.NamePrefix())
> ```

These `MayResolve` methods look up stuff by name and really shouldn't be used when you already know the ID. If the `buildCtx` and the `descs.Collection` need new methods to look up schemas by ID in a temp-schema-friendly-way, then we should go ahead and add them instead.
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go, line 302 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttemzE4gR6wk1UgWDG:-MttemzE4gR6wk1UgWDH:bfbyq2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go#L302)):*
> ```Go
> 	// CheckFeature returns if the feature name specified is allowed or disallowed,
> 	// by the database administrator.
> 	CheckFeature(ctx context.Context, schemaFeatureName string) error
> ```

It feels weird that `scbuild` should know how to transform a `tree.Statement` into a feature name, which is something that's owned by the planner. It feels like `schemaFeatureName` should have its own type like `type SchemaFeatureName string` somewhere in `tree` or somewhere high up, along with a function that takes a `tree.Statement` and turns it into a `SchemaFeatureName` by taking the two first words or whatever.
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go, line 42 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttaShf4_YCgJ9oP-Bs:-MttaShf4_YCgJ9oP-Bt:bxs0p1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go#L42)):*
> ```Go
> 			sc.SchemaKind() == catalog.SchemaTemporary {
> 			panic(pgerror.Newf(pgcode.InvalidSchemaName,
> 				"cannot drop schema %q", sc.GetName))
> ```

Shouldn't this be in `dropSchema instead`?
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go, line 54 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-Mttaa2JDh2bU5pdpfCl:-Mttaa2JDh2bU5pdpfCm:bsylqrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go#L54)):*
> ```Go
> 	sc catalog.SchemaDescriptor,
> 	behavior tree.DropBehavior,
> 	parentDrop bool,
> ```

Can you give this a better name?
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go, line 98 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttbRW18MfdUY6H6m92:-MttbRW18MfdUY6H6m93:bq0jjzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go#L98)):*
> ```Go
> 		}
> 		// For dropped types validate if any unsupported cascade is needed.
> 		for _, typeDesc := range droppedTypes {
> ```

This feels like it should be factored out. Surely this logic will come into play not only when dropping schemas?
___
*[pkg/sql/schemachanger/scdeps/exec_deps.go, line 135 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttcbPREDBsfaehs_WH:-MttcbPREDBsfaehs_WI:bt1xdrl) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scdeps/exec_deps.go#L135)):*
> ```Go
> 	}
> 	return schema.GetName(), nil
> }
> ```

Why not return the schema descriptor? Also, why do we need the `isTemporary` argument? The desc collection should do the right thing.
___
*[pkg/sql/logictest/testdata/logic_test/create_index, line 240 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttdBwdBFen9lO-UH8_:-MttdBwdBFen9lO-UH8a:btqmvmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/logictest/testdata/logic_test/create_index#L240)):*
> ```
> statement ok
> SET experimental_use_new_schema_changer = 'off'
> ```

Knowing how these things go, you'd do well to add statements to reset `experimental_use_new_schema_changer` and `jobs.debug.pausepoints` to their previous values at the end of the file, otherwise someone else is going to come along and append unrelated subtests which will unintentionally avoid the new schema changer.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 1009 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttYtq24R8-oumgnHQf:-MttYtq24R8-oumgnHQg:b20143g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/logictest/testdata/logic_test/fk#L1009)):*
> ```
> SELECT job_type FROM [SHOW JOBS] WHERE job_type = 'SCHEMA CHANGE GC' AND
>  (description LIKE 'GC for updating table "a"%' OR
>   description = 'GC for dropping descriptor 105');
> ```

Relying on the descriptor ID might end up being problematic. If possible, do something like `description = 'GC for dropping descriptor ' || <tablename>::REGCLASS::INT::STRING`
___
*[pkg/sql/schemachanger/scbuild/testdata/drop_database, line 210 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttakrN5l6kzNWa5vbV:-MttakrN5l6kzNWa5vbW:b-9peiqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/922e4986aeb2df6b3ddfd135108852eaa3584b9a/pkg/sql/schemachanger/scbuild/testdata/drop_database#L210)):*
> ```
>     dependentObjects:
>     - 55
>     - 56
> ```

Is this normal? Just checking.


<!-- Sent from Reviewable.io -->

fqazi committed:
schemachanger: use old schema changer for dropping temp schemas

Previously, we would try to clean up temporary schemas using
the declarative schema changer and run into problems looking
up schemas. This was inadequate because these temporary schemas
can only be looked up by ID within the same session. To address this,
this patch will revert to the existing schema changer for dropping
these objects. Longer term we will enhance collections and other interfaces
to help support this in the declarative schema changer.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73493)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, and @postamar)

___
*[pkg/sql/schemachanger/scbuild/descriptor_reader.go, line 39 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-Mtt_GCuBvwAPu74BHG6:-MtxbROI2GKX8JXab8Fm:bcl1rqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/922e4986aeb2df6b3ddfd135108852eaa3584b9a/pkg/sql/schemachanger/scbuild/descriptor_reader.go#L39)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

It seems strange to remove this. 
</blockquote></details>

Good point, yes I'll add it back. Additionally, for drop database, we can do a dual look up where we can try by ID first followed by schema name next as fall back (in case the name is temporary).
___
*[pkg/sql/schemachanger/scbuild/name_resolver.go, line 123 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-Mtt_Tq70BNY_DKH4vPV:-MuRBt7X703JrVhx0EhC:bndwyjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/name_resolver.go#L123)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

These `MayResolve` methods look up stuff by name and really shouldn't be used when you already know the ID. If the `buildCtx` and the `descs.Collection` need new methods to look up schemas by ID in a temp-schema-friendly-way, then we should go ahead and add them instead.
</blockquote></details>

Done.

Removed the logic added for temporary schema, we will revisit this stuff later.
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go, line 302 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttemzE4gR6wk1UgWDG:-MuRBtu9Aiy0C6aybxQZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go#L302)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

It feels weird that `scbuild` should know how to transform a `tree.Statement` into a feature name, which is something that's owned by the planner. It feels like `schemaFeatureName` should have its own type like `type SchemaFeatureName string` somewhere in `tree` or somewhere high up, along with a function that takes a `tree.Statement` and turns it into a `SchemaFeatureName` by taking the two first words or whatever.
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go, line 42 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttaShf4_YCgJ9oP-Bs:-Mtxmidi-zgBZ81aLaE9:b2oq309) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go#L42)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't this be in `dropSchema instead`?
</blockquote></details>

No, this was moved out because the same code path is used to clean underlying objects. However, I do see the danger here we should push the virtual check back in there.
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go, line 54 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-Mttaa2JDh2bU5pdpfCl:-MtxoZTbAW-LjwK5XYgb:b424yrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go#L54)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Can you give this a better name?
</blockquote></details>

Done.

Renamed to: databaseIsBeingDropped
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go, line 98 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttbRW18MfdUY6H6m92:-Mty4H1X6QRDBeZdczX7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go#L98)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This feels like it should be factored out. Surely this logic will come into play not only when dropping schemas?
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scdeps/exec_deps.go, line 135 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttcbPREDBsfaehs_WH:-MuRC07r9iak38XlcdYq:bl81tc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/schemachanger/scdeps/exec_deps.go#L135)):*
> Done.
> 
> Removed the logic added for temporary schema, we will revisit this stuff later.
Done.

Removed the logic added for temporary schema, we will revisit this stuff later.
___
*[pkg/sql/logictest/testdata/logic_test/create_index, line 240 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttdBwdBFen9lO-UH8_:-Mu1IF0V3Ib7xEpwZh3V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/logictest/testdata/logic_test/create_index#L240)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Knowing how these things go, you'd do well to add statements to reset `experimental_use_new_schema_changer` and `jobs.debug.pausepoints` to their previous values at the end of the file, otherwise someone else is going to come along and append unrelated subtests which will unintentionally avoid the new schema changer.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 1009 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttYtq24R8-oumgnHQf:-MuRMEfe7ckXe8aNLxLq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ad3c84d7153be1a1c9efeceabc1f81752e06d3f/pkg/sql/logictest/testdata/logic_test/fk#L1009)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Relying on the descriptor ID might end up being problematic. If possible, do something like `description = 'GC for dropping descriptor ' || <tablename>::REGCLASS::INT::STRING`
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scbuild/testdata/drop_database, line 210 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MttakrN5l6kzNWa5vbV:-MuRPM3O-dkmehlFoIoT:bm0mtp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/922e4986aeb2df6b3ddfd135108852eaa3584b9a/pkg/sql/schemachanger/scbuild/testdata/drop_database#L210)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this normal? Just checking.
</blockquote></details>

Done.

This was incorrect and should be gone now


<!-- Sent from Reviewable.io -->

fqazi commented:
CI will be broken until the constraint ID stuff gets merged in too. Also will refresh for active roleset stuff when its merged.

ajwerner commented on pkg/sql/logictest/logic.go:
```diff
@@ -611,6 +614,13 @@ var logicTestConfigs = []testClusterConfig{
 		overrideDistSQLMode: "on",
 		overrideAutoStats:   "false",
 	},
+	{
```
why `fakedist`? I'd really like to see a regular `local` variant with a single node.

ajwerner commented on pkg/sql/alter_type.go:
```diff
@@ -492,7 +492,7 @@ func (p *planner) canModifyType(ctx context.Context, desc *typedesc.Mutable) err
 	if !hasOwnership {
 		return pgerror.Newf(pgcode.InsufficientPrivilege,
-			"must be owner of type %s", tree.Name(desc.GetName()))
+			"must be owner of type %q", tree.Name(desc.GetName()))
```
I'd sort of prefer we kept these and change the declarative schema changer to use `%s and the `tree.Name` type which will quote identifiers it needs to quote in its string method.

ajwerner commented on pkg/sql/logictest/testdata/logic_test/drop_database:
```diff
@@ -219,19 +226,20 @@ CREATE TABLE constraint_db.t2 (
 DROP DATABASE constraint_db CASCADE
 
-query TTT
+query TT
```
how does this help us? We don't create this many jobs, do we? Should we just `skipif` this query?

ajwerner replied:
What do you think of this? Why are there two jobs in the new schema changer? It seems to me like this query is cheating because there are other jobs which meet these more lax criteria

ajwerner commented on pkg/sql/logictest/testdata/logic_test/drop_table:
```diff
@@ -31,7 +31,7 @@ DROP TABLE a
 # the startup migration are handled.
 query TT
-SELECT job_type, status FROM [SHOW JOBS] WHERE user_name = 'root' AND (job_type = 'SCHEMA CHANGE GC' OR (job_type = 'SCHEMA CHANGE' AND description != 'updating privileges'))
+SELECT replace(job_type, 'NEW SCHEMA CHANGE', 'SCHEMA CHANGE'), status FROM [SHOW JOBS] WHERE (user_name = 'root' OR user_name='node') AND (job_type = 'SCHEMA CHANGE GC' OR job_type = 'NEW SCHEMA CHANGE' OR (job_type = 'SCHEMA CHANGE' AND description != 'updating privileges'))
```
nit: wrap this

ajwerner commented on pkg/sql/logictest/testdata/logic_test/drop_type:
```diff
@@ -359,7 +359,7 @@ CREATE TYPE schema_to_drop.typ AS ENUM ('a');
 CREATE TABLE schema_to_drop.t (k schema_to_drop.typ PRIMARY KEY);
 
-statement error pgcode 0A000 unimplemented: cannot drop type "test.schema_to_drop._typ" because other objects \(\[test\.public\.t\]\) still depend on it
+statement error pgcode 0A000 unimplemented: cannot drop type "test.schema_to_drop.(_)*typ" because other objects \(\[test\.public\.t\]\) still depend on it
```
nit: `_?`

ajwerner replied:
This one is interesting, but not that interesting.

ajwerner commented on pkg/sql/logictest/testdata/logic_test/event_log:
```diff
@@ -373,12 +373,15 @@ DROP DATABASE IF EXISTS othereventlogtest CASCADE
+  (json_extract_path(info::JSONB, 'DroppedSchemaObjects') @> '["eventlogtest.public.anothertesttable", "eventlogtest.public.testtable"]' OR
+   json_extract_path(info::JSONB, 'DroppedSchemaObjects')  @> '["eventlogtest.public.testtable", "eventlogtest.public.anothertesttable"]')
 ----
-1  {"DatabaseName": "eventlogtest", "DroppedSchemaObjects": ["eventlogtest.public.anothertesttable", "eventlogtest.public.testtable"], "EventType": "drop_database", "Statement": "DROP DATABASE eventlogtest CASCADE", "Tag": "DROP DATABASE", "User": "root"}
```
Do we not do `DroppedSchemaObjects`?

ajwerner commented on pkg/sql/logictest/testdata/logic_test/rename_column:
```diff
@@ -49,7 +49,7 @@ id name  species
 user testuser
 
-statement error pq: must be owner of table users or have CREATE privilege on table users
+statement error (pq: must be owner of table users or have CREATE privilege on table users)* (user testuser does not have CREATE privilege on relation users)*
```
Is it hard to make these error messages match? Also I think the `|` would be better than the two `*`. RIght now, this will match `pq   `.

ajwerner commented on pkg/sql/logictest/testdata/logic_test/views:
```diff
@@ -1205,10 +1205,12 @@ DROP table t1nest CASCADE
 FROM system.eventlog
-WHERE "eventType" in ('drop_view', 'drop_table')
+WHERE "eventType" in ('drop_view', 'drop_table') AND
+(json_extract_path(info::JSONB, 'CascadeDroppedViews') @> '["db2.public.v3nest", "db2.public.v2nest", "db2.public.v1nest"]' OR
+ json_extract_path(info::JSONB, 'CascadeDroppedViews')  @> '["db2.public.v1nest", "db2.public.v2nest", "db2.public.v3nest"]')
```
I don't understand this, both sides of this disjunction are the same. `a @> b` says that b's elements are a subset of a's without any order. 

Maybe:
```sql
(SELECT a <@ b AND a @> b
  FROM (
        SELECT json_extract_path(info::JSONB, 'CascadeDroppedViews') AS a,
               '
                ["db2.public.v3nest", "db2.public.v2nest", "db2.public.v1nest"]
               '::JSONB AS b
       ))
```

ajwerner commented on pkg/sql/logictest/testdata/logic_test/views:
```diff
@@ -1205,10 +1205,12 @@ DROP table t1nest CASCADE
 # Validate the objects being dropped
 query IT
-SELECT "reportingID", info::JSONB - 'Timestamp' - 'DescriptorID'
+SELECT "reportingID", info::JSONB - 'Timestamp' - 'DescriptorID' - 'CascadeDroppedViews'
```
Why not show the cascaded dropped views?

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go:
```diff
@@ -30,7 +33,7 @@ func DropSchema(b BuildCtx, n *tree.DropSchema) {
 			continue
 		}
-		dropSchema(b, db, sc, n.DropBehavior)
+		dropSchema(b, db, sc, n.DropBehavior, false /* parent drop*/)
```
nit: `databaseIsBeingDropped`

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go:
```diff
@@ -40,7 +43,28 @@ func dropSchema(
+			"cannot drop schema %q", sc.GetName))
+	}
+	// This code is used for cascaded drops for databases. So, it's valid for
+	//public schemas or temporary ones to be dropped here.
```
uber-nit: space

ajwerner reviewed (commented):
Excitingly close.

fqazi committed:
schemachanger:  privilege error for schemas were quoted

The declarative schema changer was incorrectly quoting
privilege errors. To keep it consistent with the old
schema changer we will stop quoting schema names in
these errors.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73493)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, and @postamar)

___
*[pkg/sql/alter_type.go, line 495 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHXOUcfY9k6REVY-r22-495:-MuXXnAh8kNSC1LO-_yS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/alter_type.go#L495)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'd sort of prefer we kept these and change the declarative schema changer to use `%s and the `tree.Name` type which will quote identifiers it needs to quote in its string method. 
</blockquote></details>

Done.
___
*[pkg/sql/logictest/logic.go, line 617 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MqD0wVQG5VrQrIbgxVV-r22-617:-MuXOGFvBBfDClwdm35r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/logic.go#L617)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why `fakedist`? I'd really like to see a regular `local` variant with a single node. 
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go, line 54 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHWZ3iu4oJO93II-r22-54:-MuXOHtMAfG7iBIh91Xn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_schema.go#L54)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

uber-nit: space
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/drop_database, line 229 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHc9KzNq-EiilAm-r22-229:-MuXOIycCWefA6bfQ0HL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/drop_database#L229)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

how does this help us? We don't create this many jobs, do we? Should we just `skipif` this query?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/drop_type, line 362 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHZkpV8S19R61eC-r22-362:-MuXQQXMDwP8R-HrLYFa:bw3x7nz) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/drop_type#L362)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This one is interesting, but not that interesting. 
</blockquote></details>

Just an ordering difference going over the type closure when dropping a schema. Both are valid reasons to fail.
___
*[pkg/sql/logictest/testdata/logic_test/rename_column, line 52 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHc9KzNq-EiilAr-r22-52:-MuXY1ItAZFNkpPYaCG7:b-3vmhch) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/rename_column#L52)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is it hard to make these error messages match? Also I think the `|` would be better than the two `*`. RIght now, this will match `pq   `.
</blockquote></details>

There are consistency issues in the old schema changer, so we need to think about which variant makes more sense. The outlier seems to be INDEX related schema errors, which match the declarative one.



_Code quote:_
```
 testuser does not have CREATE privilege on relation users
```
___
*[pkg/sql/logictest/testdata/logic_test/views, line 1208 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHYFP3YhT4Qcgm--r22-1208:-MuXQA9Z2gM-t8dRVFct:b7maaq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/views#L1208)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Why not show the cascaded dropped views?
</blockquote></details>

The ordering differs, between the two schema changers, so I excluded it and threw it in the where clause.
___
*[pkg/sql/logictest/testdata/logic_test/views, line 1212 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHYFP3YhT4Qcgm--r22-1212:-MuXQFkdBRpOrzpZ_yVk:b-9sdpht) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/views#L1212)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't understand this, both sides of this disjunction are the same. `a @> b` says that b's elements are a subset of a's without any order. 

Maybe:
```sql
(SELECT a <@ b AND a @> b
  FROM (
        SELECT json_extract_path(info::JSONB, 'CascadeDroppedViews') AS a,
               '
                ["db2.public.v3nest", "db2.public.v2nest", "db2.public.v1nest"]
               '::JSONB AS b
       ))
``` 
</blockquote></details>

Only need to extract and compare, the @> operation doesn't care about order


<!-- Sent from Reviewable.io -->

ajwerner reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73493)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @msbutler, and @postamar)

___
*[pkg/sql/logictest/testdata/logic_test/drop_type, line 362 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHZkpV8S19R61eC-r22-362:-MuXbWk13zeY5iK7JxMc:b-w1hk2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/drop_type#L362)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Just an ordering difference going over the type closure when dropping a schema. Both are valid reasons to fail.
</blockquote></details>

yeah, hence not that interesting
___
*[pkg/sql/logictest/testdata/logic_test/views, line 1208 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHYFP3YhT4Qcgm--r22-1208:-MuX_WyW7OdsalwXHCPn:b-u9agwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/views#L1208)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

The ordering differs, between the two schema changers, so I excluded it and threw it in the where clause.
</blockquote></details>

Ah, in that case, I think my preference is to sort them in the result here and elsewhere

```sql
SELECT "reportingID",
       (info::JSONB - 'Timestamp' - 'DescriptorID')
       || ( -- sort the views for determinism
            SELECT json_build_object('CascadeDroppedViews', json_agg(value ORDER BY value))
              FROM ROWS FROM (json_array_elements((info::JSONB)->'CascadeDroppedViews'))
        )
  FROM system.eventlog
 WHERE "eventType" IN ('drop_view', 'drop_table');
```
___
*[pkg/sql/logictest/testdata/logic_test/views, line 1212 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHYFP3YhT4Qcgm--r22-1212:-MuX_Oax6QgYOIRZCpVJ:b-82ilcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/views#L1212)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Only need to extract and compare, the @> operation doesn't care about order
</blockquote></details>

See above convo


<!-- Sent from Reviewable.io -->

fqazi committed:
schemachanger: add ownership to privilege errors

Previously, when the declaratives schema changer
generated errors it only mentioned the missing
privilege. But the old schema changer also mentions
ownership inside its messages. To get the errors
consistent the declarative schema changer will
be modified to also mention ownership.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73493)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @msbutler, and @postamar)

___
*[pkg/sql/logictest/testdata/logic_test/drop_type, line 362 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHZkpV8S19R61eC-r22-362:-MuXi5fG1CSCpq71w6ND:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/drop_type#L362)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

yeah, hence not that interesting
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/logictest/testdata/logic_test/new_schema_changer:
```diff
@@ -1128,8 +1128,8 @@ SELECT "reportingID", info::JSONB - 'Timestamp' - 'DescriptorID'
 ----
-1  {"DatabaseName": "'db1-a'", "DroppedSchemaObjects": ["'db1-a'.public", "'db1-a'.sc1", "'db1-a'.sc2"], "EventType": "drop_database", "Statement": "DROP DATABASE \"'db1-a'\" CASCADE", "Tag": "DROP DATABASE", "User": "root"}
-1  {"DatabaseName": "db2", "DroppedSchemaObjects": ["db2.public", "db2.sc3"], "EventType": "drop_database", "Statement": "DROP DATABASE db2 CASCADE", "Tag": "DROP DATABASE", "User": "root"}
+1  {"DatabaseName": "'db1-a'", "DroppedSchemaObjects": ["\"'db1-a'\".public.\"'t1-esc'\""], "EventType": "drop_database", "Statement": "DROP DATABASE \"'db1-a'\" CASCADE", "Tag": "DROP DATABASE", "User": "root"}
```
why did this change?

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_table.go:
```diff
@@ -145,7 +154,8 @@ func dropTableDependents(b BuildCtx, tbl catalog.TableDescriptor, behavior tree.
 					pgcode.DependentObjectsStillExist,
-					"cannot drop table a because other objects depend on it"))
+					"cannot drop table %s because other objects depend on it",
+					tbl.GetName()))
```
tree.Name?

ajwerner commented on pkg/sql/logictest/testdata/logic_test/new_schema_changer:
```diff
@@ -457,7 +457,7 @@ CREATE TABLE defaultdb.sq1dep (
 );
 
-statement error cannot drop table a because other objects depend on it
+statement error cannot drop table shipments because other objects depend on it
```
😦 👍

ajwerner commented on pkg/sql/logictest/testdata/logic_test/views:
```diff
@@ -1205,10 +1205,20 @@ DROP table t1nest CASCADE
+ LIMIT 1;
 ----
- 1  {"CascadeDroppedViews": ["db2.public.v3nest", "db2.public.v2nest", "db2.public.v1nest"], "EventType": "drop_table", "Statement": "DROP TABLE db2.public.t1nest CASCADE", "TableName": "db2.public.t1nest", "Tag": "DROP TABLE", "User": "root"}
+1  {"CascadeDroppedViews": null, "EventType": "drop_view", "Statement": "DROP VIEW test.public.num_ref_view", "Tag": "DROP VIEW", "User": "root", "ViewName": "test.public.num_ref_view"}
```
this `null` is alarming

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73493)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, and @postamar)

___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_table.go, line 158 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHWZ3iu4oJO93IJ-r24-158:-MuYPNT81zsM7mUHrvK7:b-kb33q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf1c858053871a79775c9379d793f3037691b33/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_table.go#L158)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

tree.Name?
</blockquote></details>

Done.

Good catch
___
*[pkg/sql/logictest/testdata/logic_test/new_schema_changer, line 460 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHZkpV8S19R61eK-r24-460:-MuYPPtjEF4biQi0OwdV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf1c858053871a79775c9379d793f3037691b33/pkg/sql/logictest/testdata/logic_test/new_schema_changer#L460)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

😦 👍 
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/new_schema_changer, line 1131 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHZkpV8S19R61eK-r24-1131:-MuYQkir6iWV3IZzj_GD:b-m0k3m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf1c858053871a79775c9379d793f3037691b33/pkg/sql/logictest/testdata/logic_test/new_schema_changer#L1131)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why did this change?
</blockquote></details>

The event log entries dropped schema objects should only contain tables/views/sequences. Our implementation was wrong, so we fixed it to line up.
___
*[pkg/sql/logictest/testdata/logic_test/views, line 1212 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHYFP3YhT4Qcgm--r22-1212:-MuYPrxA-j9-ENHP2_Nm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/views#L1212)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

See above convo
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/views, line 1224 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHYFP3YhT4Qcgm--r24-1224:-MuYPsY53n8M3vozzuNA:b84n52p) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf1c858053871a79775c9379d793f3037691b33/pkg/sql/logictest/testdata/logic_test/views#L1224)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this `null` is alarming
</blockquote></details>

I messed up the sort order on the query accidentally :facepalm: 


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/logictest/logic.go:
```diff
@@ -1661,6 +1672,20 @@ func (t *logicTest) newCluster(serverArgs TestServerArgs, opts []clusterOpt) {
+			); err != nil {
+				t.Fatal(err)
+			}
+		} else {
+			if _, err := conn.Exec(
+				"SET CLUSTER SETTING sql.defaults.experimental_new_schema_changer.enabled = 'off'",
+			); err != nil {
+				t.Fatal(err)
+			}
+		}
```
It's a fresh cluster each time, so I don't think you need this

ajwerner reviewed (approved):
Address the one jobs query comment and LGTM!

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73493)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, and @postamar)

___
*[pkg/sql/logictest/logic.go, line 1688 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MqD0wVQG5VrQrIbgxVV-r26-1688:-MuaAZsVEJyj2HiUMyAB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab52a21fe4f564706a46dba28a89627476d165ce/pkg/sql/logictest/logic.go#L1688)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It's a fresh cluster each time, so I don't think you need this 
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/drop_database, line 229 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/73493#-MtkRXHc9KzNq-EiilAm-r22-229:-MuaA_Ra-9_0OXel5u5W:b-tnihd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/200beccde77bd2204144b1bd5d268ad0583a1a7e/pkg/sql/logictest/testdata/logic_test/drop_database#L229)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What do you think of this? Why are there two jobs in the new schema changer? It seems to me like this query is cheating because there are other jobs which meet these more lax criteria 
</blockquote></details>

Done.

I skipped this query for declarative and restored the old one back. I somehow missed doing this in 2 passes, I covered other queries but not the one that was commented on :facepalm: 


<!-- Sent from Reviewable.io -->

fqazi commented:
bors r+

fqazi commented:
bors r+
